### PR TITLE
Fixbug: json load

### PR DESCRIPTION
**Pull request by worker24h**
1) FixBug: load a json object is failed in routine load
    for example json-data:  {"category":"11","title":"SayingsoftheCentury","price":895,"timestamp":1589191587}

2) add `json_root` for nest json data
   for example json-data:
        {
        "RECORDS":[
            {"category":"11","title":"SayingsoftheCentury","price":895,"timestamp":1589191587},
            {"category":"22","author":"2avc","price":895,"timestamp":1589191487},
            {"category":"33","author":"3avc","title":"SayingsoftheCentury","timestamp":1589191387}
            ]
        }
Routine load:
        CREATE ROUTINE LOAD example_db.test1 ON example_tbl
        COLUMNS(category, author, price, timestamp, dt=from_unixtime(timestamp, '%Y%m%d'))
        PROPERTIES
        (
            "desired_concurrent_number"="3",
            "max_batch_interval" = "20",
            "max_batch_rows" = "300000",
            "max_batch_size" = "209715200",
            "strict_mode" = "false",
            "format" = "json",
            "jsonpaths" = "[\"$.category\",\"$.author\",\"$.price\",\"$.timestamp\"]",
            "strip_outer_array" = "true",
            "json_root" = "$.RECORDS"
        )
        FROM KAFKA
        (
            "kafka_broker_list" = "broker1:9092,broker2:9092,broker3:9092",
            "kafka_topic" = "my_topic",
            "kafka_partitions" = "0,1,2",
            "kafka_offsets" = "0,0,0"
        );

### CODE REVIEW

**Review by morningman — CHANGES_REQUESTED** [review]

File: be/src/exec/json_scanner.cpp [hunk 1]
```diff
@@ -197,30 +201,52 @@ JsonReader::~JsonReader() {
...
+Status JsonReader::init(std::string& jsonpath, std::string& json_root) {
     // parse jsonpath
+    if (!jsonpath.empty()) {
+        Status st = _generater_json_paths(jsonpath, _parsed_jsonpaths);
+        RETURN_IF_ERROR(st);
+    }
+    if (!json_root.empty()) {
+        std::vector<JsonPath> parsed_paths;
+        JsonFunctions::parse_json_paths(json_root, &parsed_paths);
+        _parsed_json_root.push_back(parsed_paths);
```
[the remark this answers is not in the document]
Comment (worker24h) [reply be/src/exec/json_scanner.cpp:213]: ok

File: be/src/exec/json_scanner.h [hunk 2]
```diff
@@ -106,33 +105,31 @@ struct JsonPath;
 class JsonReader {
 public:
     JsonReader(RuntimeState* state, ScannerCounter* counter, RuntimeProfile* profile, FileReader* file_reader,
-            std::string& jsonpath, bool strip_outer_array);
+            bool strip_outer_array);
     ~JsonReader();
 
-    Status init(); // must call before use
+    Status init(std::string& jsonpath, std::string& json_root); // must call before use
```
[the remark this answers is not in the document]
Comment (worker24h) [reply be/src/exec/json_scanner.h:111]: ok

File: be/src/exec/json_scanner.h [hunk 3]
```diff
@@ -144,9 +141,9 @@ class JsonReader {
     RuntimeProfile::Counter* _bytes_read_counter;
     RuntimeProfile::Counter* _read_timer;
     std::vector<std::vector<JsonPath>> _parsed_jsonpaths;
+    std::vector<std::vector<JsonPath>> _parsed_json_root;
     rapidjson::Document _json_doc;
-    //key: column name
-    std::unordered_map<std::string, JsonDataInternal> _jmap;
+    rapidjson::Value *_json_root;
```
[the remark this answers is not in the document]
Comment (worker24h) [reply be/src/exec/json_scanner.h:146]: ok

File: be/src/http/http_common.h [hunk 4]
```diff
@@ -37,6 +37,7 @@ static const std::string HTTP_STRICT_MODE = "strict_mode";
 static const std::string HTTP_TIMEZONE = "timezone";
 static const std::string HTTP_EXEC_MEM_LIMIT = "exec_mem_limit";
 static const std::string HTTP_EXEC_JSONPATHS  = "jsonpaths";
+static const std::string HTTP_EXEC_JSONROOT  = "json_root";
```
Comment (morningman) [inline be/src/http/http_common.h:40]:
```suggestion
static const std::string HTTP_JSONROOT  = "json_root";
```

EXEC is not some kind of prfix, it is with `EXEC_MEM_LIMIT`.
Also modify `HTTP_EXEC_JSONPATHS` to `HTTP_JSONPATHS`

File: fe/fe-core/src/main/java/org/apache/doris/load/routineload/RoutineLoadJob.java [hunk 5]
```diff
@@ -17,6 +17,14 @@
 
 package org.apache.doris.load.routineload;
 
+import com.google.common.base.Joiner;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/load/routineload/RoutineLoadJob.java:20]: Import order

File: be/src/exec/json_scanner.h [hunk 6]
```diff
@@ -143,10 +140,12 @@ class JsonReader {
...
     RuntimeProfile::Counter* _bytes_read_counter;
     RuntimeProfile::Counter* _read_timer;
+
     std::vector<std::vector<JsonPath>> _parsed_jsonpaths;
-    rapidjson::Document _json_doc;
-    //key: column name
-    std::unordered_map<std::string, JsonDataInternal> _jmap;
+    std::vector<JsonPath> _parsed_json_root;
+
+    rapidjson::Document _orinal_json_doc; // orinal json document object from parsed json string
```
[the remark this answers is not in the document]
Comment (worker24h) [reply be/src/exec/json_scanner.h]: ok

File: be/src/http/http_common.h (hunk 4 above)
Comment (worker24h) [reply be/src/http/http_common.h]: ok

File: fe/fe-core/src/main/java/org/apache/doris/load/routineload/RoutineLoadJob.java (hunk 5 above)
Comment (worker24h) [reply fe/fe-core/src/main/java/org/apache/doris/load/routineload/RoutineLoadJob.java]: ok

File: docs/en/sql-reference/sql-statements/Data Manipulation/ROUTINE LOAD.md [hunk 7]
```diff
@@ -178,6 +178,9 @@ FROM data_source
     8. `strip_outer_array`
         Boolean type, true to indicate that json data starts with an array object and flattens objects in the array object, default value is false.
 
+    9. `json_root`
+        User specifies the JSON root node as the valid JSONPath string.
```
Comment (imay) [inline docs/en/sql-reference/sql-statements/Data Manipulation/ROUTINE LOAD.md:182]: What is the json_root default value? Please let others know

File: be/src/exec/json_scanner.cpp [hunk 8]
```diff
@@ -197,30 +201,50 @@ JsonReader::~JsonReader() {
...
+        if (_strip_outer_array) {
+            _handle_json_callback = &JsonReader::_handle_flat_array_complex_json;
+        } else {
+            _handle_json_callback = &JsonReader::_handle_nested_complex_json;
+        }
+    }
+    return Status::OK();
+}
+
+Status JsonReader::_generate_json_paths(const std::string& jsonpath, std::vector<std::vector<JsonPath>>& vect) {
```
Comment (imay) [inline be/src/exec/json_scanner.cpp:227]:
```suggestion
Status JsonReader::_generate_json_paths(const std::string& jsonpath, std::vector<std::vector<JsonPath>>* vect) {
```

**Review by imay — CHANGES_REQUESTED** [review]

**Review by morningman — DISMISSED** [review]
LGTM

**Review by imay — APPROVED** [review]
LGTM

**Pull request merged**